cmd/graph.go

```diff
@@ -1,34 +1,110 @@
 package cmd
 
 import (
+	"encoding/json"
+	"html/template"
+	"net/http"
+	"sort"
+
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 	"github.com/srl-wim/container-lab/clab"
 )
 
+const (
+	templatePath = "/etc/containerlab/templates/d3js/index.html"
+)
+
+var srv string
+var tmpl string
+
+type graphTopo struct {
+	Nodes []containerDetails `json:"nodes,omitempty"`
+	Links []link             `json:"links,omitempty"`
+}
+type link struct {
+	Source         string `json:"source,omitempty"`
+	SourceEndpoint string `json:"source_endpoint,omitempty"`
+	Target         string `json:"target,omitempty"`
+	TargetEndpoint string `json:"target_endpoint,omitempty"`
+}
+
+type topoData struct {
+	Name string
+	Data template.JS
+}
+
 // graphCmd represents the graph command
 var graphCmd = &cobra.Command{
 	Use:   "graph",
 	Short: "generate a topology graph",
 
-	Run: func(cmd *cobra.Command, args []string) {
+	RunE: func(cmd *cobra.Command, args []string) error {
 		c := clab.NewContainerLab(debug)
 
 		if err := c.GetTopology(topo); err != nil {
-			log.Fatal(err)
+			return err
 		}
 
 		// Parse topology information
 		if err := c.ParseTopology(); err != nil {
-			log.Fatal(err)
+			return err
+		}
+
+		if srv == "" {
+			if err := c.GenerateGraph(topo); err != nil {
+				return err
+			}
+			return nil
+		}
+		gtopo := graphTopo{
+			Nodes: make([]containerDetails, 0, len(c.Nodes)),
+			Links: make([]link, 0, len(c.Links)),
+		}
+		for name, n := range c.Nodes {
+			gtopo.Nodes = append(gtopo.Nodes, containerDetails{
+				Name:  name,
+				Kind:  n.Kind,
+				Image: n.Image,
+				Group: n.Group,
+			})
+		}
+		sort.Slice(gtopo.Nodes, func(i, j int) bool {
+			return gtopo.Nodes[i].Name < gtopo.Nodes[j].Name
+		})
+		for _, l := range c.Links {
+			gtopo.Links = append(gtopo.Links, link{
+				Source:         l.A.Node.ShortName,
+				SourceEndpoint: l.A.EndpointName,
+				Target:         l.B.Node.ShortName,
+				TargetEndpoint: l.B.EndpointName,
+			})
+		}
+		b, err := json.Marshal(gtopo)
+		if err != nil {
+			return err
+		}
+		log.Debug(string(b))
+		topoD := topoData{
+			Name: c.Config.Name,
+			Data: template.JS(string(b)),
 		}
+		tmpl := template.Must(template.ParseFiles(tmpl))
+		http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+			tmpl.Execute(w, topoD)
+		})
 
-		if err := c.GenerateGraph(topo); err != nil {
-			log.Error(err)
+		log.Printf("Listening on %s...", srv)
+		err = http.ListenAndServe(srv, nil)
+		if err != nil {
+			return err
 		}
+		return nil
 	},
 }
 
 func init() {
 	rootCmd.AddCommand(graphCmd)
+	graphCmd.Flags().StringVarP(&srv, "srv", "", "", "server address to view, customize and download your topology PNG")
+	graphCmd.Flags().StringVarP(&tmpl, "template", "", templatePath, "golang html template used to generate the graph")
```

Comment:
add a shorthand `-t`?

I would have removed golang from description: "path to HTML template used for graph generation"

Reply:
`-t` is already taken by `--topo` in rootCmd.
why remove golang?, it's not any template

Reply:
this additional info doesn't seem important to me to explain it in the flag descr

but it might be my personal feeling on what makes a good flag descr.
If you intend to keep it, maybe change Golang->Go as the former is discouraged to use

Reply:
I think we also need to set a default path here in the flag that will point to `/etc/containerlab/templates/d3js/index.html`

Reply:
The default value is in the const `templatePath`

Reply:
updated golang to Go, and changed the const name to `defaultTemplatePath`